Print the migration file of current schema version number.

JuanitoFatas opened the pull request:
Hello there,

I use `rake db:version` to debug migration error. And the current behaviour is to print the version number of schema: 

``` Shell
$ rake db:version
Current version: 20150316153147
```

Then I use this number to find the migration file and fix the migration error.

This pull request will print the migration file path, then one can click on Terminal or copy this path to quickly open the migration file.

```
$ rake db:version
Current version: /Users/Juan/dev/rails-4-2/db/migrate/20150316153147_rename_columns_of_users.rb.
```

Hope this make  `rake db:version` a little bit better.

JuanitoFatas committed:
Add rake db:version:number and db:version:file tasks.

- rake db:version:number

  Prints the current schema version number.

- rake db:version:file

  Prints the current schema's migration file path. If the current schema version
  is ahead of the current branch an error will be raised.

claudiob commented:
I think this change can be useful, and wonder if it makes sense to take it a step further, and just print out the file name, so you can pipe it directly into your editor command.

Maybe at that point it shouldn't be called `db:version` anymore… I don't have an opinion on that.
If it replaces the current behavior, it might deserve a line in the CHANGELOG.

senny commented:
considering that rake tasks are a public facing API this looks like a breaking change. Of course this is only relevant if there are non-human consumers of `bin/rake db:version` which use the output somehow.

A [quick search](https://github.com/search?utf8=%E2%9C%93&q=%22rake+db%3Aversion%22&type=Code&ref=searchresults) on GitHub does reveal multiple uses of this command in scripts.

I wonder wether we should simply add new tasks to make the behavior more explicit and easier to process. For example:

```
$ bin/rake db:version:number
20150316153147
```

```
$ bin/rake db:version:file
/Users/Juan/dev/rails-4-2/db/migrate/20150316153147_rename_columns_of_users.rb
```

claudiob commented:
@senny I think that's a smart idea.
We could keep the current behavior for `rake db:version` and add a new rake task for the file name, so we can avoid backwards incompatibility.

JuanitoFatas commented:
I'll work on this. @senny and @claudiob Thanks for the feedback guys! :heart:

JuanitoFatas commented:
I have added these two tasks. `db:version:number` will print the current version number of schema and `db:version:file` will print the particular migration file path.

`db:version:file` has an issue is that we could be at newer version when we switched into an old branch. If that is the case, now I just make it fall back to print version number. Or should we print a message says that your migration is ahead of current branch...

What do you think?

claudiob commented on activerecord/lib/active_record/railties/databases.rake:
```diff
@@ -158,6 +158,21 @@ db_namespace = namespace :db do
 
+  namespace :version do
+    desc 'Prints the current schema version number'
+    task :number => [:environment, :load_config] do
```
@JuanitoFatas I know this is really minor but the [Rails Coding Conventions](http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#write-your-code) say:

> Use Ruby >= 1.9 syntax for hashes. Prefer { a: :b } over { :a => :b }.

Even if the rest of the file uses the old syntax, it's better to stick with the new syntax in new code.

matthewd replied:
Disagree; it's better like this.

claudiob replied:
Okay then nevermind :ribbon:

JuanitoFatas replied:
Thanks anyway :sparkles:

claudiob commented on activerecord/lib/active_record/railties/databases.rake:
```diff
@@ -158,6 +158,21 @@ db_namespace = namespace :db do
+    end
+
+    desc "Prints the file path of current schema's migration"
+    task :file => [:environment, :load_config] do
```
Here too :lipstick:

kuldeepaggarwal commented on activerecord/lib/active_record/railties/databases.rake:
```diff
@@ -158,6 +158,21 @@ db_namespace = namespace :db do
+    desc "Prints the file path of current schema's migration"
+    task :file => [:environment, :load_config] do
+      current_version = ActiveRecord::Migrator.current_version.to_s
+      migration_file = Dir.glob("#{ActiveRecord::Migrator.migrations_path}/*.rb")
```
migration_file =  Dir.glob("#{ActiveRecord::Migrator.migrations_path}/*#{ current_version  }*.rb").first

jonatack replied:
@JuanitoFatas' code is working for me and `migration_file = Dir.glob("#{ ActiveRecord::Migrator.migrations_path }/#{ current_version }.rb").first` does not.

kuldeepaggarwal replied:
Actually github parses my response in other way,, there is an asterisk ( * ) around interpolation of current_version.

* #{ current_version }

jonatack replied:
Cool, works:
```ruby
migration_file = Dir.glob("#{ActiveRecord::Migrator.migrations_path}/*#{current_version}*.rb").first
```

JuanitoFatas replied:
Thanks guys. I've applied your suggestions. :smile:

jonatack commented on activerecord/lib/active_record/railties/databases.rake:
```diff
@@ -158,6 +158,21 @@ db_namespace = namespace :db do
+      migration_file = Dir.glob("#{ActiveRecord::Migrator.migrations_path}/*.rb")
+                         .find { |migration| migration.include?(current_version) }
+      puts "Current version: #{migration_file || current_version}."
+    end
```
If I understand the usage, it might be more practical to display the filename without a trailing `.`

JuanitoFatas replied:
Thanks. I also removed the `Current version:` to make it possible to passing to pipe. :smile_cat:

rafaelfranca commented:
Could you add a CHANGELOG entry?

rafaelfranca commented:
What is the reason of `db:version:number`

JuanitoFatas commented:
> Could you add a CHANGELOG entry?

:ok_hand: Added. Please see [rendered CHANGELOG](https://github.com/JuanitoFatas/rails/blob/7f4990083663e5ccbccc3c88589fa48fcc5cb875/activerecord/CHANGELOG.md).


> What is the reason of db:version:number

I was trying to modify existing `db:version` rake task. But `db:version` has [non-human consumers](https://github.com/search?utf8=%E2%9C%93&q=%22rake+db%3Aversion%22&type=Code&ref=searchresults), so instead of modifying the `db:version` task, [@senny suggests that](https://github.com/rails/rails/pull/19367#issuecomment-82912323) adds two explicit rake task, one is to print the current version number of schema, which people can use in their script in order to be explicit; The second one is the purpose of this Pull Request which prints the file path of migration file. :blush:

jonatack commented:
@JuanitoFatas for the last sentence in the changelog, I'd suggest something like:

"Prints the current schema's migration file path, or, if the current schema version is ahead of the current branch and no migration file can be found, the current version number."

or

"Prints the current schema's migration file path. If the current schema version is ahead of the current branch and no migration file can be found, falls back to printing the current version number."

JuanitoFatas commented:
@jonatack Thanks. Changed to `Prints the current schema's migration file path. If the current schema version is ahead of the current branch and no migration file can be found, it falls back to printing the current version number.`.

matthewd commented:
> it falls back to printing the current version number

I think this should be an error condition. If the invoking script/application/whatever is obliged to work out whether we just gave it a filename or a useless number, it might as well just do the glob search itself.

matthewd commented on activerecord/lib/active_record/railties/databases.rake:
```diff
@@ -158,6 +158,20 @@ db_namespace = namespace :db do
+    desc "Prints the file path of current schema's migration"
+    task :file => [:environment, :load_config] do
+      current_version = ActiveRecord::Migrator.current_version
+      migration_file = Dir.glob("#{ActiveRecord::Migrator.migrations_path}/*#{current_version}*.rb").first
```
Instead of `/*#{current_version}*.rb`, can't we use `/#{current_version}_*.rb`? It seems improbable to have a migration-number-looking-thing in the middle of a migration's filename, but not impossible.

JuanitoFatas replied:
Makes sense, I changed to `/#{current_version}_*.rb`, thanks!

JuanitoFatas commented:
> I think this should be an error condition. If the invoking script/application/whatever is obliged to work out whether we just gave it a filename or a useless number, it might as well just do the glob search itself.

What do you think about the error message? What could we help the user to find out their current migration is ahead of the current branch?

Thanks!

jonatack commented:
I like your idea of a warning message.

In that case, for the change log you could write something like:

"Prints the current migration file path, or, if the current version is ahead of the current branch, raises (or prints, depending on what you decide to do here) a warning message."

senny commented:
@JuanitoFatas can you add test-cases? I think they are located in the railties test suite.

JuanitoFatas commented:
> @JuanitoFatas can you add test-cases? I think they are located in the railties test suite.

Ok. Will work on it.
